File: adt2unit/i386/calc_freq_shift_up.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

function calc_freq_shift_up(freq,shift: Word): Word;

var
  result: Word;

begin
  asm
        mov     cx,freq
        mov     ax,shift
        mov     bx,cx
        and     bx,0000001111111111b
        mov     dx,cx
        and     dx,0001110000000000b
        add     bx,ax
        and     cx,1110000000000000b
        shr     dx,10
        cmp     bx,FreqEnd
        jb      @@2
        cmp     dx,7
        jnz     @@1
        mov     bx,FreqEnd
        jmp     @@2
@@1:    sub     bx,FreqRange
        inc     dx
@@2:    mov     ax,cx
        shl     dx,10
        add     ax,dx
        add     ax,bx
        mov     result,ax
  end;
  calc_freq_shift_up := result;
end;

File: adt2unit/calc_freq_shift_up.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

/*static*/ uint16_t calc_freq_shift_up (uint16_t freq, uint16_t shift) {
  uint16_t a, b, c;

  a = (freq & 0x03FF/*00000011 11111111*/) + shift;
  b =  freq & 0x1C00/*00011100 00000000*/;
  c =  freq & 0xE000/*11100000 00000000*/;

  if (a >= FreqEnd) { // FIXME: isn't it supposed to be `>' here?
    if (b != (7 << 10)) {
      a -= FreqRange;
      b += 1 << 10;
    } else
      a = FreqEnd;
  }

  return a + b + c;
}
